src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4222,11 +4175,11 @@ private static void MarkSlotsAsNotNull(ArrayBuilder<int> slots, ref LocalState s
 
         private void LearnFromNonNullTest(BoundExpression expression, ref LocalState state)
         {
-            if (expression.Kind == BoundKind.AwaitableValuePlaceholder)
+            if (expression is BoundValuePlaceholderBase placeholder)
             {
-                if (_awaitablePlaceholdersOpt != null && _awaitablePlaceholdersOpt.TryGetValue((BoundAwaitableValuePlaceholder)expression, out var value))
+                if (_resultForPlaceholdersOpt != null && _resultForPlaceholdersOpt.TryGetValue(placeholder, out var value))
```

Comment:
nit: Thanks for renaming the map to something more general. Along those lines, we may want to add a helper methods for placeholder replacement (here and in `VisitAwaitExpression`)